my-account/app/myaccount/login/LogInPageController.java

```diff
@@ -35,66 +36,65 @@
-            return completedFuture(handleLogInFormErrors(boundForm, pageContent, userContext));
+        final Form<LogInFormData> logInForm = logInUnboundForm.bindFromRequest();
+        if (logInForm.hasErrors()) {
+            return handleLogInFormErrors(logInForm, userContext);
         } else {
-            return logIn(boundForm.get())
-                    .thenApplyAsync(signInResult -> handleSuccessfulSignIn(signInResult, userContext), HttpExecution.defaultContext())
-                    .exceptionally(throwable -> handleInvalidCredentialsError(throwable, pageContent, userContext));
+            return logIn(logInForm.get())
+                    .thenComposeAsync(signInResult -> handleSuccessfulSignIn(signInResult, userContext), HttpExecution.defaultContext())
+                    .exceptionally(throwable -> handleInvalidCredentialsError(throwable, logInForm, userContext).toCompletableFuture().join()); // TODO move to async
         }
     }
 
     @RequireCSRFCheck
     public CompletionStage<Result> processSignUp(final String languageTag) {
         final UserContext userContext = userContext(languageTag);
-        final Form<SignUpFormData> boundForm = signUpForm.bindFromRequest();
-        final LogInPageContent pageContent = new LogInPageContent(boundForm, userContext, i18nResolver(), configuration());
-        if (boundForm.hasErrors()) {
-            return completedFuture(handleSignUpFormErrors(boundForm, pageContent, userContext));
+        final Form<SignUpFormData> signUpForm = signUpUnboundForm.bindFromRequest();
+        if (signUpForm.hasErrors()) {
+            return handleSignUpFormErrors(signUpForm, userContext);
         } else {
-            return signUp(boundForm.get())
-                    .thenApplyAsync(signInResult -> handleSuccessfulSignIn(signInResult, userContext), HttpExecution.defaultContext())
-                    .exceptionally(throwable -> handleExistingCustomerError(throwable, pageContent, userContext));
+            return signUp(signUpForm.get())
+                    .thenComposeAsync(signInResult -> handleSuccessfulSignIn(signInResult, userContext), HttpExecution.defaultContext())
+                    .exceptionally(throwable -> handleExistingCustomerError(throwable, signUpForm, userContext).toCompletableFuture().join());  // TODO move to async
         }
     }
 
-    public Result processLogOut(final String languageTag) {
+    public CompletionStage<Result> processLogOut(final String languageTag) {
         CustomerSessionUtils.removeCustomer(session());
         CartSessionUtils.removeCart(session());
-        return redirect(reverseRouter().showHome(languageTag));
+        final Call call = reverseRouter().showHome(languageTag);
+        return completedFuture(redirect(call));
     }
 
-    private CompletionStage<CustomerSignInResult> logIn(final LogInFormData formData) {
+    protected CompletionStage<CustomerSignInResult> logIn(final LogInFormData formData) {
```

Comment:
Yes, this will be handled when converting Sunrise into a framework, step by step:
https://github.com/sphereio/commercetools-sunrise-java/issues/368